lib/Service/ImageService.php

```diff
@@ -0,0 +1,631 @@
+	 * @param string $url
+	 * @param $resource
+	 * @return array
+	 */
+	private function simpleDownload(string $url, $resource): array {
+		$client = $this->clientService->newClient();
+		try {
+			$options = [
+				// does not work with sink if SSE is enabled
+				// 'sink' => $resource,
+				// rather use stream and write to the file ourselves
+				'stream' => true,
+				'timeout' => 0,
+				'headers' => [
+					'User-Agent' => 'Nextcloud Text',
+				],
+			];
+
+			$response = $client->get($url, $options);
+			$body = $response->getBody();
+			while (!feof($body)) {
+				// write ~5 MB chunks
+				$chunk = fread($body, 5000000);
+				fwrite($resource, $chunk);
+			}
+
+			return ['Content-Type' => $response->getHeader('Content-Type')];
+		} catch (ServerException | ClientException $e) {
+			//$response = $e->getResponse();
+			//if ($response->getStatusCode() === 401) {
+			$this->logger->warning('Impossible to download image', ['exception' => $e]);
+			return ['error' => 'Impossible to download image'];
+		} catch (ConnectException $e) {
+			$this->logger->error('Connection error', ['exception' => $e]);
+			return ['error' => 'Connection error'];
+		} catch (Throwable | Exception $e) {
+			$this->logger->error('Unknown download error', ['exception' => $e]);
+			return ['error' => 'Unknown download error'];
+		}
+	}
+
+	/**
+	 * Actually delete attachment files which are not pointed in the markdown content
+	 *
+	 * @param int $fileId
+	 * @return int The number of deleted files
+	 * @throws NotFoundException
+	 * @throws NotPermittedException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Lock\LockedException
+	 * @throws \OC\User\NoUserException
+	 */
+	public function cleanupAttachments(int $fileId): int {
+		$textFile = $this->rootFolder->getById($fileId);
+		if (count($textFile) > 0 && $textFile[0] instanceof File) {
+			$textFile = $textFile[0];
+			if ($textFile->getMimeType() === 'text/markdown') {
+				// get IDs of the files inside the attachment dir
+				try {
+					$attachmentDir = $this->getAttachmentDirectoryForFile($textFile);
+				} catch (NotFoundException $e) {
+					// this only happens if the attachment dir was deleted by the user while editing the document
+					return 0;
+				}
+				$attachmentsByName = [];
+				foreach ($attachmentDir->getDirectoryListing() as $attNode) {
+					$attachmentsByName[$attNode->getName()] = $attNode;
+				}
+
+				$contentAttachmentNames = $this->getAttachmentNamesFromContent($textFile->getContent());
+
+				$toDelete = array_diff(array_keys($attachmentsByName), $contentAttachmentNames);
+				foreach ($toDelete as $name) {
+					$attachmentsByName[$name]->delete();
+				}
+				return count($toDelete);
+			}
+		}
+		return 0;
+	}
+
+
+	/**
+	 * Get attachment file names listed in the markdown file content
+	 *
+	 * @param string $content
+	 * @return array
+	 */
+	private function getAttachmentNamesFromContent(string $content): array {
```

Comment:
Bonus points for some quick unit test ;)

Reply:
Yep, will do after fixing smaller stuff.

Reply:
A small test was added in c351e66607d23b1672603752b2bdd1dba83b59a6
Do you have more ideas of potentially problematic names?

Reply:
Looks good :)